### ISSUE TITLE
[help wanted] debugging broken autocomplete in Windows

### ISSUE THREAD

**Issue by darikg** [jonathanslenders/python-prompt-toolkit#119]
Autocomplete has been broken for me in Windows (Windows 7 x64) for a while now, but I've just had time recently to try to investigate. When using `python-prompt-toolkit/examples/autocompletion.py`, pressing tab doesn't show the autocomplete menu, and pressing control-tab actually hangs the prompt.

By throwing logging statements everywhere, I think I've got the problem narrowed down pretty specifically, but I'm not sure where to go from here. The issue seems to localize to `prompt-toolkit/eventloop/win32._wait_for_handles`. When given an array of handles, `windll.kernel32.WaitForMultipleObjects` returns `-1`, which is presumably an error code, but it just gets used as an index into the `handles` array, and so it always returns the stdin handle, instead of raising an error.

Here's a minimal example:

```
from ctypes import windll, c_long, pointer
from ctypes.wintypes import DWORD, BOOL
from prompt_toolkit.win32_types import STD_INPUT_HANDLE, SECURITY_ATTRIBUTES

h1 = windll.kernel32.GetStdHandle(STD_INPUT_HANDLE)
h2 = windll.kernel32.CreateEventA(pointer(SECURITY_ATTRIBUTES()), BOOL(True), BOOL(False), None)


def my_wait_func(handles):
    arrtype = c_long * len(handles)
    handle_array = arrtype(*handles)
    timeout = 100
    handle_idx = windll.kernel32.WaitForMultipleObjects(len(handle_array), handle_array, False, DWORD(timeout))
    err_code = windll.kernel32.GetLastError()
    return handle_idx, err_code


print(my_wait_func([h1]))           # prints (0, 0) [Ok]
print(my_wait_func([h2]))           # prints (258, 0) (timeout) [Ok]
print(my_wait_func([h1, h2]))       # prints (-1, 6) [BAD!]
print(my_wait_func([h1]))           # prints (0, 0) [Still ok]
print(my_wait_func([h2]))           # prints (258, 0) [Still ok]
```

According to [this](https://msdn.microsoft.com/en-us/library/windows/desktop/ms681382(v=vs.85).aspx), an error code of 6 corresponds to  ERROR_INVALID_HANDLE, but I don't understand why it works fine when called with a scalar array.

**Comment by jonathanslenders** [comment]
Hi @darikg,

Amazing that you got so far debugging the issue. Return code -1 is indeed an issue.
According to this page: https://msdn.microsoft.com/en-us/library/windows/desktop/ms687025%28v=vs.85%29.aspx it's actually not -1, but 0xFFFFFFFF. I guess that's the same.

Now, the weird thing is that on my Windows laptop. (Also Windows 7, 64 bit), for both Python 2 and Python 3. This script seems to work fine:
```
C:\Users\jslender\Downloads>\Python27\python.exe  test.py
(0, 0)
(258, 0)
(0, 0)
(0, 0)
(258, 0)

C:\Users\jslender\Downloads>\Python34\python.exe  test.py
(0, 0)
(258, 0)
(0, 0)
(0, 0)
(258, 0)
```

When running the code inside Idle however, it doesn't: (Python 3.4.3) I think because stdin is not a real input in my case.
```
(-1, 6)
(258, 0)
(-1, 6)
(-1, 6)
(258, 0)
```

To give a bit more context. The reason for using ``WaitForMultipleObjects``, is to have a blocking call that waits until there is either user input *or* an executor (a thread) calls ``call_from_executor``. This last call puts a callback from the thread in a queue, and sets the ``Event`` object. This way, the main thread can go on and run that piece of code.

Autocompletion is done asynchronously, so in a separate thread. But when the result comes back, it is shown by using ``call_from_executor``.

Now, handling the -1 error code is easy. I could just ignore it and call ``WaitForMultipleObjects`` again. But I'm afraid that results in an eternal loop. That's not what we want.

Could you try changing ``c_long`` with ``DWORD``? According to the MSISDN doc, it expects a WORD. (WORD looks to be a ``c_ulong``, not ``c_long``.) Hopefully, that works. If not, we have to try something else.

Jonathan

**Comment by darikg** [comment]
Thanks @jonathanslenders! Through sheer trial & error, I got it to work, by using the `HANDLE` type:

```
from ctypes.wintypes import DWORD, BOOL, HANDLE
# [...]
    arrtype = HANDLE * len(handles)
```
(Whether the False was wrapped in a BOOL or not didn't seem to matter.) When you get a chance, could you try the same on your machine?

**Comment by jonathanslenders** [comment]
Hi thanks @darikg . This makes me happy. I will try this tonight.

What do you get when executing the following code?
```python
>>> from ctypes.wintypes import HANDLE
>>> HANDLE
<class 'ctypes.c_void_p'>
>>>
```
HANDLE is a ``c_void_p`` as well ?

What do you get when executing the following line?
```
import sys;print("%x" % sys.maxsize, sys.maxsize > 2**32)
```
I get: ```7fffffff False```. It means that I'm 32 bit. Maybe you have 64 bit Python which caused void pointers to have another size.

Jonathan

**Comment by darikg** [comment]
I think you're right!

```
[pgcli2] C:\Users\dg\Documents\Python\pgcli>python
Python 2.7.10 |Continuum Analytics, Inc.| (default, May 28 2015, 16:44:52) [MSC v.1500 64 bit (AMD64)] on win32

Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://binstar.org
>>> from ctypes.wintypes import HANDLE
>>> HANDLE
<class 'ctypes.c_void_p'>
>>> import sys;print("%x" % sys.maxsize, sys.maxsize > 2**32)
('7fffffffffffffff', True)
>>>
```

**Comment by jonathanslenders** [comment]
Hi @darikg,

I was also able to reproduce the issue on Python 3.5, 64bit.
It has been fixed in this commit: https://github.com/jonathanslenders/python-prompt-toolkit/commit/f26ca1ee03fefcd1a992608cec6ac4f9bb499875

A bit more testing and I'll release again.
Thanks a lot with all the help fixing this issue!

Jonathan

**Comment by darikg** [comment]
That's awesome. It even fixes jonathanslenders/ptpython#30
